File: csharp/902.numbers-at-most-n-given-digit-set.cs
/*
 * @lc app=leetcode id=902 lang=csharp
 *
 * [902] Numbers At Most N Given Digit Set
 */

// @lc code=start
public class Solution {
    public int AtMostNGivenDigitSet(string[] D, int N) {
        string s = N.ToString();
        int len = s.Length, ret = 0, k = D.Length;
        int[] pow = new int[len];
        pow[0] = 1;
        for (int i = 1; i < len; ++i) {
            pow[i] = pow[i - 1] * k;
            ret += pow[i];
        }
        for (int i = 0; i < len; ++i) {
            char c = s[i];
            bool end = true;
            foreach (string d in D) {
                if (d[0] < c) {
                    ret += pow[len - i - 1];
                } else {
                    if (d[0] == c) {
                        end = false;
                    }
                    break;
                }
            }
            if (end) {
                return ret;
            }
        }
        return ret + 1;
    }
}
// @lc code=end

File: java/902.numbers-at-most-n-given-digit-set.java
/*
 * @lc app=leetcode id=902 lang=java
 *
 * [902] Numbers At Most N Given Digit Set
 */

// @lc code=start
class Solution {
    public int atMostNGivenDigitSet(String[] digits, int n) {
        String s = Integer.toString(n);
        int k = s.length(), ret = 0, m = digits.length;
        int[] pow = new int[k];
        pow[0] = 1;
        for (int i = 1; i < k; ++i) {
            pow[i] = m * pow[i - 1];
            ret += pow[i];
        }
        for (int i = 0; i < k; ++i) {
            char t = s.charAt(i);
            boolean flag = false;
            for (String d : digits) {
                char c = d.charAt(0);
                if (c < t) {
                    ret += pow[k - 1 - i];
                } else {
                    if (c == t) {
                        flag = true;
                    }
                    break;
                }
            }
            if (!flag) {
                return ret;
            }
        }
        return ret + 1;
    }
}
// @lc code=end

